### PR TITLE
Conform metrics' names to OpenTelemetry spec

### PR DESCRIPTION
**Pull request by heyams**
#37880 is in Harsima's fork. To make it easier for myself, i replaced it with this PR.

### CODE REVIEW

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java [hunk 1]
```diff
@@ -149,9 +151,17 @@ public static void updateMetricPointBuilder(
...
 
         pointBuilder.setValue(pointDataValue);
-        // TODO (heya) why give it the same name as otel metric?
-        //  it seems this field doesn't matter and only _MS.MetricId property matters?
-        pointBuilder.setName(metricData.getName());
+
+        // We emit some metrics via OpenTelemetry that have names which use characters that aren't
+        // supported in OpenTelemetry metric names, and so we put the real metric names into an attribute
+        // (where these characters are supported) and then pull the name back out when sending it to Breeze.
+        String metricName = pointData.getAttributes().get(APPLICATIONINSIGHTS_INTERNAL_METRIC_NAME);
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java:158]: I think we also want to skip this attribute when emitting it to Breeze

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java [hunk 2]
```diff
@@ -45,6 +46,7 @@ public class MetricDataMapper {
 
     private static final Set<String> OTEL_PRE_AGGREGATED_STANDARD_METRIC_NAMES = new HashSet<>(4);
     private static final List<String> EXCLUDED_METRIC_NAMES = new ArrayList<>();
+    public static final AttributeKey<String> APPLICATIONINSIGHTS_INTERNAL_METRIC_NAME = AttributeKey.stringKey("applicationinsights_internal_metric_name");
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java:49]: (follows attribute naming conventions a bit more)
```suggestion
    public static final AttributeKey<String> APPLICATIONINSIGHTS_INTERNAL_METRIC_NAME = AttributeKey.stringKey("applicationinsights.internal.metric_name");
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java (hunk 1 above)
Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java:158]: updated the comment

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java [hunk 3]
```diff
@@ -164,7 +165,10 @@ public static void updateMetricPointBuilder(
 
         metricTelemetryBuilder.setMetricPoint(pointBuilder);
 
+        // remove any attributes that start with "applicationinsights.internal."
         Attributes attributes = pointData.getAttributes();
+        attributes.asMap().keySet().removeIf(it -> it.getKey().startsWith("applicationinsights.internal."));
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java:170]: I think you were right about this already being excluded

Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java]: we were using EMPTY_MAPPING and didn't ignore the standard metrics attribute prefixes.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java [hunk 4]
```diff
@@ -186,7 +195,8 @@ public static void updateMetricPointBuilder(
                     metricTelemetryBuilder, statusCode, success, dependencyType, target, isSynthetic);
             }
         } else {
-            EMPTY_MAPPINGS.map(attributes, metricTelemetryBuilder);
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java:189]: here is the bug.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java (hunk 3 above)
Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/MetricDataMapper.java]: do we want to ingore all these prefixes for metrics: 

 "http.",
                "db.",
                "message.",
                "messaging.",
                "rpc.",
                "enduser.",
                "net.",
                "peer.",
                "exception.",
                "thread.",
                "faas.",
                "code.",
                "job.", // proposed semantic convention which we use for job,
                "applicationinsights.internal."));

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorMetricExporterTest.java [hunk 5]
```diff
@@ -139,8 +139,8 @@ public void testAttributesOnCustomMetric() {
         assertThat(metricsData.getMetrics().size()).isEqualTo(1);
         Map<String, String> properties = metricsData.getProperties();
 
-        assertThat(properties.size()).isEqualTo(2);
-        assertThat(properties.get(NET_PEER_NAME.getKey())).isEqualTo("example.io");
+        assertThat(properties.size()).isEqualTo(1);
+        assertThat(properties.get(NET_PEER_NAME.getKey())).isNull();
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorMetricExporterTest.java:143]: "net.peer.name" is part of the to be ignored standard metrics prefixes.

**Review by jeanbisutti — APPROVED** [review]

**Pull request merged**